### PR TITLE
hashi_vault: restore compatibility with <=v2.9

### PR DESCRIPTION
**Pull request by flowerysong**
#### SUMMARY

As noted in https://github.com/ansible-collections/community.general/pull/23#issuecomment-612984880, https://github.com/ansible/ansible/pull/64288 changed the data returned by the `hashi_vault` lookup in a backwards-incompatible way. It would be nice to fix this before the next stable release and avoid introducing a regression.

This PR retains the new ability to request a specific field or all field values for KV v2, but returns the same structure that Ansible v2.9 does in other cases.

I also reworked the tests slightly, mainly making sure we test all of the permutations of return values for both KV versions and trying to specify the return values as tightly as possible to prevent future
regressions.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

hashi_vault

##### ADDITIONAL INFORMATION
stable-2.9:
```
TASK [No field specified] ******************************************************
ok: [localhost] => 
  lookup('hashi_vault', 'secret=moresecrets/data/foo:'):
    data:
      field1: hay
      field2: clover
    metadata:
      created_time: '2020-05-06T05:33:27.650170871Z'
      deletion_time: ''
      destroyed: false
      version: 1

TASK [data field] **************************************************************
ok: [localhost] => 
  lookup('hashi_vault', 'secret=moresecrets/data/foo:data'):
    field1: hay
    field2: clover

TASK [real field] **************************************************************
fatal: [localhost]: FAILED! => 
  msg: 'An unhandled exception occurred while running the lookup plugin ''hashi_vault''. Error was a <class ''ansible.errors.AnsibleError''>, original message: The secret moresecrets/data/foo does not contain the field ''field1''. for hashi_vault lookup'
...ignoring
```

Current state:
```
TASK [No field specified] ******************************************************
ok: [localhost] => 
  lookup('community.general.hashi_vault', 'secret=moresecrets/data/foo:'):
    field1: hay
    field2: clover

TASK [data field] **************************************************************
fatal: [localhost]: FAILED! => 
  msg: 'An unhandled exception occurred while running the lookup plugin ''community.general.hashi_vault''. Error was a <class ''ansible.errors.AnsibleError''>, original message: The secret moresecrets/data/foo does not contain the field ''data''. for hashi_vault lookup'
...ignoring

TASK [real field] **************************************************************
ok: [localhost] => 
  lookup('community.general.hashi_vault', 'secret=moresecrets/data/foo:field1'): hay
```

After this change:
```
TASK [No field specified] ******************************************************
ok: [localhost] => 
  lookup('community.general.hashi_vault', 'secret=moresecrets/data/foo:'):
    data:
      field1: hay
      field2: clover
    metadata:
      created_time: '2020-05-06T05:33:27.650170871Z'
      deletion_time: ''
      destroyed: false
      version: 1

TASK [data field] **************************************************************
ok: [localhost] => 
  lookup('community.general.hashi_vault', 'secret=moresecrets/data/foo:data'):
    field1: hay
    field2: clover

TASK [real field] **************************************************************
ok: [localhost] => 
  lookup('community.general.hashi_vault', 'secret=moresecrets/data/foo:field1'): hay
```

### CODE REVIEW

**Comment by felixfontein** [comment]
CC @onitake @briantist @ihard @Im0 (all involved in the other PR)

**Comment by felixfontein** [comment]
@flowerysong does changelogs/fragments/64288-fix-hashi-vault-kv-v2.yaml need to be adjusted?

**Comment by flowerysong** [comment]
> @flowerysong does changelogs/fragments/64288-fix-hashi-vault-kv-v2.yaml need to be adjusted?

Definitely. I'm not even sure what that entry is trying to say; none of the changes (at least in the PR as merged) have any effect on what version of the secret is returned.

File: plugins/lookup/hashi_vault.py [hunk 1]
```diff
@@ -335,28 +335,27 @@ def get(self):
...
 
-        if field not in data['data']:
-            raise AnsibleError("The secret %s does not contain the field '%s'. for hashi_vault lookup" % (secret, field))
+        # a single field was requested
+        if field in data['data']:
+            return [data['data'][field]]
+
+        # handle nested data for KV v2
+        if is_kv_v2 and field in data['data']['data']:
+            return [data['data']['data'][field]]
```
Comment (briantist) [inline plugins/lookup/hashi_vault.py:356]: Since the effective result of the changes is akin to changing the default value of `return_format` to `raw`, in this section you could check for `return_format == 'dict'`, and if that's set then do the v2 nested return first; otherwise use 2.9 behavior. That would allow for a v2 secet with a `data` field to be returned as expected for someone aware of how to request it.

File: plugins/lookup/hashi_vault.py [hunk 2]
```diff
@@ -335,28 +335,27 @@ def get(self):
...
         # https://vaultproject.io/api/secret/kv/kv-v2.html#sample-response-1
-        try:
-            # sentinel field checks
-            check_dd = data['data']['data']
-            check_md = data['data']['metadata']
-            # unwrap nested data
-            data = data['data']
-        except KeyError:
-            pass
+        is_kv_v2 = '/data/' in secret and set(data['data'].keys()) == set(('data', 'metadata'))
```
Comment (briantist) [inline plugins/lookup/hashi_vault.py:339]: Very nice!

**Comment by flowerysong** [comment]
> I think the first example is unintuitive. The value after the colon is meant to be a field, but there is no field named `data`; the second example keeps the original (though v1 I guess?) meaning of what field is.

This behaviour is the price of remaining compatible with something that has grown organically rather than being designed from the ground up for the current state. Speaking as someone who unexpectedly had to update dozens of variables the last time someone changed the lookup's behaviour to be more "intuitive", I'm very wary of claims that people finding something unintuitive warrants breaking other people's playbooks.

It also matches the behaviour of the `vault` CLI tool:
```
$ vault read -field=data moresecrets/data/foo
map[data:inafieldofyellowflowers field1:hay field3:sojabohne]
```

> What happens if the secret has an actual field called `data`?

You have to request the entire secret and access it via Jinja, just like in 2.9.

(Arguably this is the only thing the lookup should support, since manipulation of the returned data is more flexible and more intuitively aligned with other parts of Ansible if you just use Jinja.)

> One thing that should be addressed is the `return_format` option I added.
> That's going to need some changes since its possible values and default were based on the behavior when I came to it.

The minimum necessary changes should already be addressed, though I do feel the `dict` format name could do with some rethinking (since it sometimes returns a dict and sometimes doesn't.)

> I added `raw` as an option that would return what 2.9 previously did, and called the post 2.9 behavior `dict`.

`raw` is not the same as 2.9, it's the entire API result (which has even more metadata and a more deeply nested structure):
```
ok: [localhost] => 
  lookup('community.general.hashi_vault', 'secret=moresecrets/data/foo', return_format='raw'):
    auth: null
    data:
      data:
        data: inafieldofyellowflowers
        field1: hay
        field3: sojabohne
      metadata:
        created_time: '2020-05-07T03:35:03.505099077Z'
        deletion_time: ''
        destroyed: false
        version: 4
    lease_duration: 0
    lease_id: ''
    renewable: false
    request_id: 3a2acae6-7bc0-6197-1695-70d6780aa8a5
    warnings: null
    wrap_info: null
```

> If going with your changes, `raw` would be the new sensible default (which would keep 2.9 compat for most examples).

Nope. `raw` is useful because it exposes data that was previously unobtainable via the lookup, but precisely because of that it's completely different from the prior default behaviour.

```
TASK [KV v1 raw] ******************************************************
ok: [localhost] => 
  lookup('community.general.hashi_vault', 'secret=secret/ping:value', return_format='raw'):
    auth: null
    data:
      value: pong
    lease_duration: 2764800
    lease_id: ''
    renewable: false
    request_id: a9a2c6a7-4353-67b1-8f19-03f12277229c
    warnings: null
    wrap_info: null
```

> tl;dr: I'm on board with changing the default back to showing `raw` output to maintain compatibility, but I think the usage of `:data` as proposed is ambiguous and should either be changed now or deprecated.

The ambiguity was introduced by people trying to add special handling for KV v2 without requiring a flag to enable the new behaviour. I'm open to rethinking this approach.

**Comment by briantist** [comment]
> > I think the first example is unintuitive. The value after the colon is meant to be a field, but there is no field named `data`; the second example keeps the original (though v1 I guess?) meaning of what field is.
> 
> This behaviour is the price of remaining compatible with something that has grown organically rather than being designed from the ground up for the current state. Speaking as someone who unexpectedly had to update dozens of variables the last time someone changed the lookup's behaviour to be more "intuitive", I'm very wary of claims that people finding something unintuitive warrants breaking other people's playbooks.

Fair points; I mostly agree except on the fact that a single specific value (`data`) shouldn't return a different result than any other value, or require the user to handle it differently, there should be consistency.

I understand how we got to this place, and the importance of maintaining a stable interface so as not to break playbooks like you've experienced, but also think this is something that warrants a breaking change (not right away mind you, and not the same breaking change you're looking to revert).

So I think we should do 2 things:

1. Use deprecation: let it be backward compatible now as you're proposing, set a deprecation, warn users it will change eventually.
1. Provide _forward compatibility_ toward the future option. This is where `return_format` can help. If not `raw` let's ensure that A) a value exists that describes the current behavior and is today's default value, and B) a value exists that describes the future behavior, and will become default after deprecation.

This path doesn't just provide a warning about when something will break, it allows users to preemptively update their playbooks during the deprecation period to an explicit `return_format` of their choice (I want the original behavior now and after deprecation, or I want the post-deprecation behavior now and after), so that the process is decoupled from upgrading and can be done over time.

(after deprecation, users who chose it explicitly could go back to having it implicitly if they want, but that can be done at their pace as well)

> > What happens if the secret has an actual field called `data`?
> 
> You have to request the entire secret and access it via Jinja, just like in 2.9.
> 
> (Arguably this is the only thing the lookup should support, since manipulation of the returned data is more flexible and more intuitively aligned with other parts of Ansible if you just use Jinja.)

I don't disagree, but that ship sailed when the `:field` mechanism was put in. Changing that now would be more disruptive imo, but I'm not opposed to that change either if it has a sufficient deprecation period and forward compat option like above that takes out some of the pain of making the change.

> > One thing that should be addressed is the `return_format` option I added.
> > That's going to need some changes since its possible values and default were based on the behavior when I came to it.
> 
> The minimum necessary changes should already be addressed, though I do feel the `dict` format name could do with some rethinking (since it sometimes returns a dict and sometimes doesn't.)

Since none of those format options are released yet I'm completely open to better names, changes to behavior and the options presented. Happy to hear any ideas you have about that.

> > I added `raw` as an option that would return what 2.9 previously did, and called the post 2.9 behavior `dict`.
> 
> `raw` is not the same as 2.9, it's the entire API result (which has even more metadata and a more deeply nested structure):
> 
> > If going with your changes, `raw` would be the new sensible default (which would keep 2.9 compat for most examples).
> 
> Nope. `raw` is useful because it exposes data that was previously unobtainable via the lookup, but precisely because of that it's completely different from the prior default behaviour.

Yeah that's completely correct; thank you. I did most of this work months ago, motivated by a project that is on hold for another quarter, so I've actually not run this stuff much since writing it! 😅
 
> > tl;dr: I'm on board with changing the default back to showing `raw` output to maintain compatibility, but I think the usage of `:data` as proposed is ambiguous and should either be changed now or deprecated.
> 
> The ambiguity was introduced by people trying to add special handling for KV v2 without requiring a flag to enable the new behaviour. I'm open to rethinking this approach.

The ambiguity I'm talking about is that of what adding `:data` does to the behavior. Maybe overloading is a better term.

I fully understand the pain of breaking changes, especially when you have to chose _at the point of upgrade_ because there's no option that works for both old and new. Let me know what you think about the deprecation path, especially the part that eases the transition pain by decoupling upgrade from playbook changes.

**Comment by felixfontein** [comment]
@flowerysong @briantist so how should we continue with this?
